### PR TITLE
SONARJAVA-3024 Cookie domain should have at least 2 levels

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/CookieDomainCheck.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      MethodMatcher.create().typeDefinition("javax.servlet.http.Cookie").name("setDomain").addParameter("java.lang.String"),
+      MethodMatcher.create().typeDefinition("java.net.HttpCookie").name("setDomain").addParameter("java.lang.String")
+    );
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    ExpressionTree arg = mit.arguments().get(0);
+    String domain = ConstantUtils.resolveAsStringConstant(arg);
+    if (domain != null && !domain.substring(1).contains(".")) {
```
Comment (christophe-zurn-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/CookieDomainCheck.java]: Check will crash if code contains a call to `setDomain` method with empty string `myCookie.setDomain("");`

File: java-checks/src/test/files/checks/CookieDomainCheck.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import javax.servlet.http.Cookie;
+import java.net.HttpCookie;
+class A {
+ void foo() {
+   Cookie myCookie = new Cookie("name", "val");
+   myCookie.setDomain(".com"); // Noncompliant [[sc=23;ec=29]] {{Specify at least a second-level cookie domain.}}
+   myCookie.setDomain(".myDomain.com"); // Compliant
+ }
+ void HttpCookie(String domain) {
+   HttpCookie myCookie = new Cookie("name", "val");
```
Comment (christophe-zurn-sonarsource) [inline java-checks/src/test/files/checks/CookieDomainCheck.java]: Should it not be `new HttpCookie` instead?

File: java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S3331_java.html [hunk 3]
```diff
@@ -0,0 +1,25 @@
+<p>A cookie's domain specifies which websites should be able to read it. Left blank, browsers are supposed to only send the cookie to sites that
```
Comment (christophe-zurn-sonarsource) [inline java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S3331_java.html]: Not directly related to implementation, but I think it's a shame that `HttpCookie` is mentioned nowhere in RSPEC description, not even as an example

File: java-checks/src/main/java/org/sonar/java/checks/CookieDomainCheck.java (hunk 1 above)
Comment (benzonico) [reply java-checks/src/main/java/org/sonar/java/checks/CookieDomainCheck.java]: nice catch !

File: java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S3331_java.html (hunk 3 above)
Comment (benzonico) [reply java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S3331_java.html]: I'll update the example in RSPEC.

**Pull request merged**